Refactor PostProcessingComponent into reactive registry of effects

MbfloydIR opened the pull request:
## Summary
Added a Highlight system to add the highlight component to selected items.
Based off of the branch michael/IR-594-postprocessing since it also effects the post processing changes


## Subtasks Checklist

## Breaking Changes

## References
closes #_insert number here_

## QA Steps

MbfloydIR committed:
added color average effect

MbfloydIR committed:
added the outline effect to the post processing

MbfloydIR committed:
added smaa effect for post processing

MbfloydIR committed:
added ChromaticAberrationEffect to post processing

MbfloydIR committed:
added DotScreenEffect post processing effect

MbfloydIR committed:
added TiltShiftEffect post processing effect

MbfloydIR committed:
Added motion blur post processing effect

MbfloydIR committed:
added Glitch Effect to post processing

MbfloydIR committed:
added lineWidth GodRaysEffect and GridEffect to post processing

MbfloydIR committed:
added LUT1DEffect to post processing

MbfloydIR committed:
added LUT3DEffect post processing effect

MbfloydIR committed:
added LUT1DEffect post processing effect

MbfloydIR committed:
added NoiseEffect to post processing

MbfloydIR committed:
added the PixelationEffect to post processing

MbfloydIR committed:
added ScanlineEffect to post processing effects

MbfloydIR committed:
added FXAAEffect and ShockWaveEffect to post processing

MbfloydIR committed:
added TextureEffect to post processing effects

MbfloydIR committed:
added LensDistortionEffect to post processing

MbfloydIR committed:
Merge branch 'dev' into michael/postprocessing

* dev: (42 commits)
  fix(magic-link): Fix guest identity provider removal upon login token validation and retain only the latest generated token (#10078)
  Support GLTF Authoring (#10088)
  fix getNestedChildren not snubing children (#10097)
  Ir 507 Refactor InteractableComponent and expose API (#10064)
  Remove AssetDatabaseType from Asset schema (#10095)
  Added Asset Database Type in schema (#10094)
  Updated argument of install-projects.js script
  Fix distance model deserialization having wrong type check (#10091)
  Add GLTF Scene Support (#10045)
  FileThumbnailJobState now unescapes file names and replaces whitespace characters with hyphens to avoid issues with escaped characters in thumbnail URLs not loading.
  Made changes to input and select components (#10081)
  Fixed a couple bugs in cronjob updating. (#10087)
  tailwind error view and other tailwind fixes (#10076)
  Improve mac os onboarding, disable minio, and update mediasoup (#10071)
  Break Up and Type GLTF Loader (#10044)
  update default scene ground plane to be a nicer colour
  Added check for aws rolearn rollback.
  Added `helm repo update` before `helm upgrade` commands (#10084)
  Added batch invalidation fields to builder helm ugprade command (#10083)
  Modal fixes & UI improvements (#10077)
  ...

# Conflicts:
#	packages/spatial/src/renderer/effects/PostProcessing.ts
#	packages/spatial/src/renderer/functions/configureEffectComposer.ts

HexaField committed:
Merge branch 'dev' into michael/IR-594-postprocessing

MbfloydIR committed:
corrected error with duplicate case option

MbfloydIR committed:
moved the effects composer code into the post processing component so we can use texture loading hooks

MbfloydIR committed:
committing latest progress for the post processing effects

MbfloydIR committed:
updated the post processing editor to use the lut path

MbfloydIR committed:
Merge branch 'dev' into michael/IR-594-postprocessing

* dev:
  fix admin projects and build status modals and statuses (#10116)
  Add unique constraint to project name column (#10125)
  9928 optimize asset builds (#9965)
  Physics API Abstraction (#10124)
  rework scene loaded state usage (#10127)
  Camera and Spectation are Event Sourced (#10130)
  Location GLTF State Cleanup (#10129)
  Add locationType menu (#10128)
  IR 465 ECS Materials (#10109)
  Fix scenestate references in playmode tool (#10126)
  Remove old ColliderComponent (#10122)
  GLTF ECS Loader (#10102)
  Systems without any pre/post/sub systems will no longer start out expanded (#10120)
  Updated the avatar radius and camera rotation (#10118)
  Fixed setup helm in dev-deploy (#10092)
  Fix interactable cursor style reactor (#10117)
  IR-1790 making the addComponent search box gain focus automatically when opening the addComponent popup (#10106)
  Update AddEditLocationModal.tsx (#10110)
  Update README.md (#10112)
  MountPoint/Grabbable/Link Components now add Interactables to the authoring layer, optimized useEffects to Systems/Components/Editors (#10105)

MichaelEstes committed:
example

MbfloydIR committed:
moved the effect logic into individual use effects

MbfloydIR committed:
added effect options conditional check

MbfloydIR committed:
removed the associative array map to the effects class since it is not being used in a for loop any longer

MbfloydIR committed:
connected the post processing effects up to the individual useEffect

MbfloydIR committed:
removed the effectMap refs that are no longer needed

MbfloydIR committed:
updated default values for post processing effects
commented out SSGIEffect since it required the composure before it is available

MbfloydIR committed:
created child reactor to wait for the useScene to return a render Entity

MbfloydIR committed:
moving composure declaration so it can be used in an effect constructor for SSREffect

MbfloydIR committed:
alphabetized the effects in the editor

MbfloydIR committed:
added missing SSGIEffect ref

DanielBelmes committed:
fix render pass

MbfloydIR committed:
added highlights to editor items that are selected

MbfloydIR committed:
Added highlight system to add highlights to selected objects

MbfloydIR committed:
Merge branch 'michael/IR-594-postprocessing' into michael/IR-1983-Highlight

* michael/IR-594-postprocessing: (34 commits)
  fix render pass
  added missing SSGIEffect ref
  alphabetized the effects in the editor
  moving composure declaration so it can be used in an effect constructor for SSREffect
  created child reactor to wait for the useScene to return a render Entity
  updated default values for post processing effects commented out SSGIEffect since it required the composure before it is available
  removed the effectMap refs that are no longer needed
  connected the post processing effects up to the individual useEffect
  removed the associative array map to the effects class since it is not being used in a for loop any longer
  added effect options conditional check
  moved the effect logic into individual use effects
  example
  updated the post processing editor to use the lut path
  committing latest progress for the post processing effects
  moved the effects composer code into the post processing component so we can use texture loading hooks
  corrected error with duplicate case option
  added LensDistortionEffect to post processing
  added TextureEffect to post processing effects
  added FXAAEffect and ShockWaveEffect to post processing
  added ScanlineEffect to post processing effects
  ...

MbfloydIR committed:
Merge branch 'dev' into michael/IR-1983-Highlight

* dev:
  IR-1282-Track-code-generated-assets-in-the-resource-manager (#10111)
  Refactoring MathConstants.ts Vector3 and folding in Axis3D.ts (#10137)
  feat(IR-200): adding required behavior to the modal to hide the cancel action (#10141)
  Admin search fixes (#10142)
  Revert change to scene save asset url (#10143)

MbfloydIR committed:
implemented the outline effect for the highlight selection
implemented the smaa default effect
removed the outline and smaa effects from the post processing editor options since they are always on

MbfloydIR committed:
updated the outline effect ref

MbfloydIR committed:
put the highlight query map ref back to correct typescript issue

DanielBelmes committed:
Merge branch 'dev' into michael/IR-1983-Highlight

MbfloydIR committed:
initial attempt at unity test

MbfloydIR committed:
updated the unity test to loop through the effects of the post processing and set its various properties using a master data obj

DanielBelmes committed:
Merge branch 'dev' into michael/IR-198-Unit-Tests-PostProcessing

DanielBelmes committed:
Move postprocessing

MbfloydIR committed:
added unit test for the effect composer

HexaField commented on packages/spatial/src/renderer/components/PostProcessingComponent.tsx:
```diff
@@ -59,16 +104,532 @@ export const PostProcessingComponent = defineComponent({
+  const postprocessingComponent = useComponent(entity, PostProcessingComponent)
+  const camera = useComponent(rendererEntity, CameraComponent)
+  const renderer = useComponent(rendererEntity, RendererComponent)
+  const composer = new EffectComposer(renderer.value.renderer)
```
this is creating a new composer and render pass every time the renderer reactor re-runs.

we should use a factory function for the initial state of this reactor like we do starting line 152 with the new scene

HexaField commented on packages/editor/src/systems/HighlightSystem.ts:
```diff
@@ -0,0 +1,55 @@
+const reactor = () => {
+  const selectedEntities = SelectionState.useSelectedEntities()
+
+  for (const entity of highlightQuery()) removeComponent(entity, HighlightComponent)
```
why are we running a query in the body of a reactor? this is an anti pattern. if the intention is to clean up the highlight component, we can do it by returning a cleanup function in the use effect

MbfloydIR replied:
query removed, clean up now done in reaction to schema.isActive being changed to false

HexaField commented on packages/spatial/src/renderer/components/PostProcessingComponent.tsx:
```diff
@@ -59,16 +104,532 @@ export const PostProcessingComponent = defineComponent({
+    velocityDepthNormalPass.set(new VelocityDepthNormalPass(scene, camera))
+  }, [scene])
+
+  useEffect(() => {
```
i think a cleaner pattern for specific effects would be to define a dictionary of effects to create/destroy functions that are invoked, rather than hardcoding each effect. we can then use sub-reactors to map each effect into it's own react component, and handle create/destroy/update events in a controlled and generic way.

MbfloydIR committed:
Updated the composure ref to use a factory function for the initial state of this reactor, so it stops creating a new composer and render pass every time the renderer reactor re-runs

MbfloydIR committed:
removed the query in the body of a reactor
removal of the highlight component, is now down by returning a cleanup function in the use effect

MbfloydIR committed:
updated highlight state and play mode connection to rely on the data of the selection state

MbfloydIR committed:
removed obsolete commented out code

MbfloydIR committed:
updated the post processesing component to not have a use effect for each effect

MbfloydIR committed:
corrected effect schema issue

MbfloydIR committed:
removed dev logs

MbfloydIR committed:
Merge branch 'michael/IR-198-Unit-Tests-PostProcessing' into michael/IR-1983-Highlight

* michael/IR-198-Unit-Tests-PostProcessing:
  removed dev logs
  corrected effect schema issue
  updated the post processesing component to not have a use effect for each effect
  removed obsolete commented out code
  updated highlight state and play mode connection to rely on the data of the selection state
  removed the query in the body of a reactor removal of the highlight component, is now down by returning a cleanup function in the use effect
  Updated the composure ref to use a factory function for the initial state of this reactor, so it stops creating a new composer and render pass every time the renderer reactor re-runs
  added unit test for the effect composer
  Move postprocessing
  fix linkedin icon on user table (#10164)
  Adds editor and reviewer project permission types and removes user (#10171)
  Remove unnecessary code for client setting seed (#10168)
  remove unused file
  Add unmount to remove component if extension is deleted (#10167)
  updated the unity test to loop through the effects of the post processing and set its various properties using a master data obj
  [IR-1828] fix: Reparenting Transform applied based on the new Parent Transform (#10161)
  initial attempt at unity test

MbfloydIR committed:
updated the post processing to avoid adding extra passes

MbfloydIR committed:
removed dev logs

HexaField commented on packages/engine/src/postprocessing/components/PostProcessingComponent.tsx:
```diff
@@ -0,0 +1,572 @@
+    depthDownsamplingPass,
+    composer
+  } = props
+  const lookUp = {
```
this mapping can be a hyperflux state or at least a module scope object (like the ones in PostProcessing.ts) such that it can be extensible

HexaField commented on packages/spatial/tsconfig.json:
```diff
@@ -41,5 +41,5 @@
     "./**/*.ts",
     "./**/*.tsx"
-  ]
+, "../engine/src/postprocessing/PostProcessing.ts"  ]
```
vscode likes to do this when moving files, please fix

MbfloydIR committed:
removed unwanted entry from the tsconfig

MbfloydIR committed:
Merge branch 'dev' into michael/IR-1983-Highlight

* dev: (63 commits)
  hotfix vrm0 data not being where we expect it (#10236)
  Update tween.js version (#10241)
  Added ProjectPermissionDatabaseType
  physics bug fix (#10221)
  updated the background to use the color with when in wireframe render mode (#10231)
  vrm expressionmanager was being thrown away it is actually needed for viseme support to work (#10230)
  IR-1887-Asset-Preview-breaking-if-you-click-gltf-that-is-in-the-scene (#10208)
  query function fix (#10229)
  IR-2102 Material/Plugin Parameters (#10180)
  Fixed webcam light not turning off when camera paused. (#10224)
  Made client's server.js not bound to a specific host (#10223)
  Update FeathersHooks.tsx (#10228)
  Updated app name
  feat(setup): update logo and components props (#10204)
  Changes for feature flag schema to be string enum (#10225)
  Cleaned location hooks (#10216)
  Ir 1652 interactable input refactor (#10219)
  IR-2018 Refactor the Select component to disable search functionality for the dropdown only (#10206)
  refactor: Update Primus initialization to include pathname in server URL (#10205)
  Move physics enter/exit back to reactors (#10193)
  ...

# Conflicts:
#	packages/editor/src/components/element/ElementList.tsx
#	packages/editor/src/components/properties/PostProcessingSettingsEditor.tsx
#	packages/engine/src/scene/SceneModule.ts
#	packages/spatial/src/renderer/components/PostProcessingComponent.tsx
#	packages/spatial/src/renderer/functions/configureEffectComposer.ts

MbfloydIR committed:
Merge branch 'dev' into michael/IR-1983-Highlight

* dev:
  URL resolution to "__$project__" on scene save
  IR-1652 input sink fixes (#10233)

MbfloydIR committed:
corrected lint issues

MbfloydIR committed:
removed configure effect composure since it is no longer used

MbfloydIR committed:
updated to fix lint errors

MbfloydIR committed:
corrected the lint error

DanielBelmes commented on packages/engine/src/postprocessing/components/PostProcessingComponent.tsx:
```diff
@@ -0,0 +1,572 @@
+    }
+  }, [schema.isActive])
+
+  useEffect(() => {
```
For the next changes we can probably just have schema.isActive as the dep and both create and destroy in one useEffect. I don't wanna break functionality if this works for now 😅

DanielBelmes commented on packages/engine/src/postprocessing/components/PostProcessingComponent.tsx:
```diff
@@ -0,0 +1,572 @@
+    const smaaPreset = getState(RenderSettingsState).smaaPreset
+    const smaaEffect = new SMAAEffect({
+      preset: smaaPreset,
+      edgeDetectionMode: EdgeDetectionMode.COLOR
```
seems like we forgot to add this to the effect composer. Gonna do it now

DanielBelmes replied:
Forgot the reason for this. A lot of the new effects are convolution effects that can't mix with SMAA. Gonna just expose in editor and allow people to decide at scene authoring I think. Though we gotta consider error messaging or handling for this library. It's a headache.

DanielBelmes committed:
Add SMAAEffect to default

HexaField reviewed (changes requested):
We spoke on discord to again change the implementation to a full blown registry, making the component extensible and well scoped.

```ts
/** Spatial Package */
type EffectRegistryEntry = {
  reactor: (props: { parameters; isActive }) => void
  defaultValues: any // specifies the default values for the effect
  schema: any // specifies a schema for the editor to generate UI for each effect
}

export const PostProcessingEffectState = defineState({
  name: 'PostProcessingEffectState',
  initial: {} as Record<string, EffectRegistryEntry>
})

// PostProcessingComponent
const reactor = () => {
  const entity = useEntityContext()
  const postProcessingComponent = useComponent(entity, PostProcessingComponent)
  const rendererEntity = useScene(entity) // get renderer entity
  const effects = useComponent(entity, PostProcessingEffectState)

  useEffect(() => {
    // create composer etc
  }, [])

  // for each effect specified in our postProcessingComponent, we mount a sub-reactor based on the effect registry for that effect ID
  return (
    <>
      {Object.entries(postProcessingComponent.effects.value).map(([key, value]) => {
        const effect = effects[key] // get effect registry entry
        return (
          <effect.reactor
            key={key}
            parameters={postProcessingComponent.effects[key]}
            isActive={value.isActive}
            rendererEntity={rendererEntity}
          />
        )
      })}
    </>
  )
}

/** Engine package */

// for effects using assets, the effect needs to exist in the engine package
const ChromaticAberrationEffectProcessReactor = (props: { parameters; isActive; rendererEntity }) => {
  const postProcessgComponent = useComponent(entity, PostProcessingComponent)
  const isLutPassActive = postProcessgComponent.effects[lutpass].isActive // if we want an effect to be active only if another effect is active
  const effectComposer = useComponent(props.rendererEntity, RendererComponent).effectComposer
  const mytexture = useAsset()

  useEffect(() => {
    if (!props.isActive || isLutPassActive) return
    const { schema, effects } = props
    const eff = new ChromaticAberrationEffect(schema)
    effects[Effects.ChromaticAberrationEffect].set(eff)
    effectComposer.addPass(eff)
    return () => {
      //cleanup
    }
  }, [props.isActive, isLutPassActive])
}

// registers the effect
getMutableState(PostProcessingEffectState)[Effects.ChromaticAberrationEffect].set({
  reactor: ChromaticAberrationEffectProcessReactor,
  defaultValues: {
    //...
  },
  schema: {
    hue: { propertyType: PropertyTypes.Number, name: 'Hue', min: -1, max: 1, step: 0.01 },
    saturation: { propertyType: PropertyTypes.Number, name: 'Saturation', min: -1, max: 1, step: 0.01 }
  }
})





/**
 * 1. create registry and move effect definitions to use registry 
 *    (removing Effects, EffectMap, defaultPostProcessingSchema etc in PostProcessing.ts)
 * 
 * 2. register each effect into the registry
 * 
 * 3. refactor PostProcessingComponent to use the registry definitions and mount sub-reactors for each effect
 * 
 * 4. refactor studio editor to use the registry definition's schemas for generating UI
 */
```

DanielBelmes committed:
Merge branch 'dev' into michael/IR-1983-Highlight

DanielBelmes committed:
I fixed badly
